encoder.go

```diff
@@ -0,0 +1,162 @@
+	defaultDivisor = 362033331456891249 // 57^10
+)
+
+func maxPow(b uint64) (d uint64, n int) {
+	d, n = b, 1
+	for m := math.MaxUint64 / b; d <= m; {
+		d *= b
+		n++
+	}
+	return
+}
+
+// Encode encodes uuid.UUID into a string using the most significant bits (MSB)
+// first according to the alphabet.
+func (e encoder) Encode(u uuid.UUID) string {
+	if e.alphabet.singleBytes {
+		return e.encodeSingleBytes(u)
+	}
+	return e.encode(u)
+}
+
+func (e encoder) encodeSingleBytes(u uuid.UUID) string {
+	num := uint128{
+		binary.BigEndian.Uint64(u[8:]),
+		binary.BigEndian.Uint64(u[:8]),
+	}
+	var r uint64
+	var i int
+	var buf []byte
+	if e.alphabet.len == defaultBase { // compiler optimizations using constants for default base
+		buf = make([]byte, defaultEncLen)
+		for i = defaultEncLen - 1; num.Hi > 0 || num.Lo > 0; {
+			num, r = num.quoRem64(defaultDivisor)
+			for j := 0; j < defaultNDigits && i >= 0; j++ {
+				buf[i] = byte(e.alphabet.chars[r%defaultBase])
+				r /= defaultBase
+				i--
+			}
+		}
+	} else {
+		buf = make([]byte, e.alphabet.encLen)
+		l := uint64(e.alphabet.len)
+		d, n := maxPow(l)
+		for i = int(e.alphabet.encLen - 1); num.Hi > 0 || num.Lo > 0; {
+			num, r = num.quoRem64(d)
+			for j := 0; j < n && i >= 0; j++ {
+				buf[i] = byte(e.alphabet.chars[r%l])
+				r /= l
+				i--
+			}
+		}
+	}
+	for ; i >= 0; i-- {
+		buf[i] = byte(e.alphabet.chars[0])
+	}
+	return string(buf[:])
+}
+
+func (e encoder) encode(u uuid.UUID) string {
+	num := uint128{
+		binary.BigEndian.Uint64(u[8:]),
+		binary.BigEndian.Uint64(u[:8]),
+	}
+	var r uint64
+	var outIndexes []uint64
+	if e.alphabet.len == defaultBase { // compiler optimizations using constants for default base
+		outIndexes = make([]uint64, defaultEncLen) // avoids escaping to heap for base57 when used with constant
+		for i := defaultEncLen - 1; num.Hi > 0 || num.Lo > 0; {
+			num, r = num.quoRem64(defaultDivisor)
+			for j := 0; j < defaultNDigits && i >= 0; j++ {
+				outIndexes[i] = r % defaultBase
+				r /= defaultBase
+				i--
+			}
+		}
+	} else {
+		outIndexes = make([]uint64, e.alphabet.encLen)
+		l := uint64(e.alphabet.len)
+		d, n := maxPow(l)
+		for i := int(e.alphabet.encLen - 1); num.Hi > 0 || num.Lo > 0; {
+			num, r = num.quoRem64(d)
+			for j := 0; j < n && i >= 0; j++ {
+				outIndexes[i] = r % l
+				r /= l
+				i--
+			}
+		}
+	}
+
+	var sb strings.Builder
+	sb.Grow(int(e.alphabet.encLen))
+	for i := 0; i < int(e.alphabet.encLen); i++ {
+		sb.WriteRune(e.alphabet.chars[outIndexes[i]])
+	}
+	return sb.String()
+}
+
+// Decode decodes a string according to the alphabet into a uuid.UUID. If s is
+// too short, its most significant bits (MSB) will be padded with 0 (zero).
+func (e encoder) Decode(s string) (u uuid.UUID, err error) {
+	var n uint128
+	var index int64
+
+	for _, char := range s {
+		index, err = e.alphabet.Index(char)
+		if err != nil {
+			return
+		}
+		if e.alphabet.len == defaultBase { // compiler optimization using constant for default base
```

Comment:
Same here, is it worth it?

Reply:
on `Decode` it's around ~11% performance difference.
You can see full benchmarks in comment above.

Reply:
I just ran benchmarks on `Decode` again, and it seems like there was some fluke in the first run.
And it seems like removing this `if` actually improves performance by ~2%.
Removed it.